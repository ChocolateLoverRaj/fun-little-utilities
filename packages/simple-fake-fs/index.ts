export enum ErrorKind {
  ENOENT = -2,
  EEXIST = -17,
  ENOTDIR = -20,
  EISDIR = -21,
}

export type ErrorCode = string & keyof typeof ErrorKind

export type SysCall =
  | 'stat'
  | 'scandir'
  | 'open'
  | 'read'
  | 'mkdir'
  | 'exists'

interface ErrorConstructor<Return, Path> {
  new (syscall: SysCall, path: Path): Return
}

export class ENOENT<Path> {
  public readonly errno = ErrorKind.ENOENT
  public readonly code = ErrorKind[ErrorKind.ENOENT]
  constructor(
    public readonly syscall: SysCall,
    public readonly path: Path,
  ) {}
}

export class EEXIST<Path> {
  public readonly errno = ErrorKind.EEXIST
  public readonly code = ErrorKind[ErrorKind.EEXIST]
  constructor(
    public readonly syscall: SysCall,
    public readonly path: Path,
  ) {}
}

export class ENOTDIR<Path> {
  public readonly errno = ErrorKind.ENOTDIR
  public readonly code = ErrorKind[ErrorKind.ENOTDIR]
  constructor(
    public readonly syscall: SysCall,
    public readonly path: Path,
  ) {}
}

export class EISDIR<Path> {
  public readonly errno = ErrorKind.EISDIR
  public readonly code = ErrorKind[ErrorKind.EISDIR]
  constructor(
    public readonly syscall: SysCall,
    public readonly path: Path,
  ) {}
}

export enum ContentKind {
  File = 'File',
  Directory = 'Directory',
  None = 'None',
  Error = 'Error',
}

const EMPTY_CLASS = class {}

class None {
  public readonly kind = ContentKind.None
}

const NONE = new None()

class ErrorCarrier<Value> {
  public readonly kind = ContentKind.Error
  constructor(public readonly value: Value) {}
}

export class FakeStats<FS extends Content<any, any>> {
  constructor(private readonly fs: FS) {}
  public readonly isFile = () => this.fs.kind === ContentKind.File
  public readonly isDirectory = () => this.fs.kind === ContentKind.Directory
}

export class FakeFileContent<Content> {
  public readonly kind = ContentKind.File
  constructor(public readonly content: Content) {}
}

interface FileSystemEntry<PathElm, FileContent> extends ReadonlyArray<any> {
  readonly 0: PathElm
  readonly 1: Content<PathElm, FileContent> | readonly FileSystemEntry<PathElm, FileContent>[]
}

export class FakeDirectoryContent<PathElm, FileContent> extends Map<
  PathElm,
  Content<PathElm, FileContent>,
> {
  public readonly kind = ContentKind.Directory

  constructor(entries: readonly FileSystemEntry<PathElm, FileContent>[] = []) {
    super(entries.map(([key, value]) => {
      if (value instanceof FakeFileContent || value instanceof FakeDirectoryContent) {
        return [key, value]
      }

      const newValue = new FakeDirectoryContent(value)
      return [key, newValue] as const
    }))
  }

  public hasPath(path: readonly PathElm[]) {
    const { kind } = this.getPath(path, 'exists', EMPTY_CLASS, EMPTY_CLASS)
    return kind === ContentKind.File || kind === ContentKind.Directory
  }

  public getPath<ERP, EFD>(
    path: readonly PathElm[],
    syscall: SysCall,
    RedundantPath: ErrorConstructor<ERP, PathElm[]>,
    FileAsDir: ErrorConstructor<EFD, PathElm[]>,
    original = () => Array.from(path),
  ): MaybeContent<
    PathElm,
    FileContent,
    ERP | EFD
  > {
    if (!path.length) return this
    const [first, ...rest] = path
    const next = this.get(first)
    if (!next) {
      return rest.length ? new ErrorCarrier(new RedundantPath(syscall, original())) : NONE
    }
    if (next.kind !== ContentKind.Directory) {
      return rest.length ? new ErrorCarrier(new FileAsDir(syscall, original())) : next
    }
    return next.getPath(rest, syscall, RedundantPath, FileAsDir, original)
  }

  public setPath<ENE, EFD>(
    path: readonly PathElm[],
    value: Content<PathElm, FileContent>,
    syscall: SysCall,
    NotExist: ErrorConstructor<ENE, PathElm[]>,
    FileAsDir: ErrorConstructor<EFD, PathElm[]>,
    original = () => Array.from(path),
  ): ENE | EFD | null {
    const [first, ...rest] = path
    if (rest.length) {
      const child = this.get(first)
      if (!child) return new NotExist(syscall, original())
      if (child.kind !== ContentKind.Directory) return new FileAsDir(syscall, original())
      return child.setPath(rest, value, syscall, NotExist, FileAsDir, original)
    }
    this.set(first, value)
    return null
  }

  public ensurePath<EFD>(
    path: readonly PathElm[],
    value: Content<PathElm, FileContent>,
    syscall: SysCall,
    FileAsDir: ErrorConstructor<EFD, PathElm[]>,
    original = () => Array.from(path),
  ): EFD | null {
    const [first, ...rest] = path
    if (rest.length) {
      const child = this.get(first)
      if (!child) {
        const next = new FakeDirectoryContent<PathElm, FileContent>()
        const error = next.ensurePath(rest, value, syscall, FileAsDir, original)
        if (error) return error
        this.set(first, next)
        return null
      }
      if (child.kind !== ContentKind.Directory) {
        return new FileAsDir(syscall, original())
      }
      return child.ensurePath(rest, value, syscall, FileAsDir, original)
    }
    this.set(first, value)
    return null
  }
}

export type Content<PathElm, FileContent> =
  | FakeFileContent<FileContent>
  | FakeDirectoryContent<PathElm, FileContent>

type MaybeContent<PathElm, FileContent, ErrorValue> =
  | None
  | ErrorCarrier<ErrorValue>
  | Content<PathElm, FileContent>

export class ArrayPathFileSystem<PathElm, FileContent> {
  private coreMap: FakeDirectoryContent<PathElm, FileContent>

  constructor(entries?: readonly FileSystemEntry<PathElm, FileContent>[]) {
    this.coreMap = new FakeDirectoryContent(entries)
  }

  public existsSync = (path: readonly PathElm[]) => this.coreMap.hasPath(path)

  public statSync = (path: readonly PathElm[]) => {
    const content = this.coreMap.getPath(path, 'stat', ENOENT, ENOTDIR)
    switch (content.kind) {
      case ContentKind.None:
        throw new ENOENT('stat', path)
      case ContentKind.Error:
        throw content.value
    }
    return new FakeStats(content)
  }

  public readdirSync = (dirname: readonly PathElm[]) => {
    const content = this.coreMap.getPath(dirname, 'scandir', ENOENT, ENOTDIR)
    switch (content.kind) {
      case ContentKind.Directory:
        return Array.from(content.keys())
      case ContentKind.File:
        throw new ENOTDIR('scandir', dirname)
      case ContentKind.None:
        throw new ENOENT('scandir', dirname)
      case ContentKind.Error:
        throw content.value
    }
  }

  public readFileSync = (filename: readonly PathElm[]) => {
    const content = this.coreMap.getPath(filename, 'read', ENOENT, ENOTDIR)
    switch (content.kind) {
      case ContentKind.File:
        return content.content
      case ContentKind.Directory:
        throw new EISDIR('read', filename)
      case ContentKind.None:
        throw new ENOENT('read', filename)
      case ContentKind.Error:
        throw content.value
    }
  }

  public mkdirSync = (dirname: readonly PathElm[]) => {
    switch (this.coreMap.getPath(dirname, 'mkdir', EMPTY_CLASS, EMPTY_CLASS).kind) {
      case ContentKind.File:
      case ContentKind.Directory:
        throw new EEXIST('mkdir', dirname)
    }
    const error = this.coreMap.setPath(dirname, new FakeDirectoryContent(), 'mkdir', ENOENT, ENOTDIR)
    if (error) throw error
  }

  public writeFileSync = (filename: readonly PathElm[], fileContent: FileContent) => {
    if (this.coreMap.getPath(filename, 'open', EMPTY_CLASS, EMPTY_CLASS).kind === ContentKind.Directory) {
      throw new EISDIR('open', filename)
    }
    const error = this.coreMap.setPath(filename, new FakeFileContent(fileContent), 'open', ENOENT, ENOTDIR)
    if (error) throw error
  }

  public ensureDirSync = (dirname: readonly PathElm[]) => {
    switch (this.coreMap.getPath(dirname, 'mkdir', ENOENT, ENOTDIR).kind) {
      case ContentKind.File:
        throw new ENOTDIR('mkdir', dirname)
      case ContentKind.Directory:
        return
    }
    const error = this.coreMap.ensurePath(dirname, new FakeDirectoryContent(), 'mkdir', ENOTDIR)
    if (error) throw error
  }

  public outputFileSync = (filename: readonly PathElm[], content: FileContent) => {
    if (this.coreMap.getPath(filename, 'open', ENOENT, ENOTDIR).kind === ContentKind.Directory) {
      throw new EISDIR('open', filename)
    }
    const error = this.coreMap.ensurePath(filename, new FakeFileContent(content), 'open', ENOTDIR)
    if (error) throw error
  }
}

type StringPathFileSystemDict = {
  readonly [_: string]: StringPathFileSystemDict | string
}

function dict2entries(dict: StringPathFileSystemDict): FileSystemEntry<string, string>[] {
  return Object.entries(dict).map(([key, value]) => {
    if (typeof value === 'string') {
      return [key, new FakeFileContent(value)]
    }

    return [key, dict2entries(value)]
  })
}

export class StringPathFileSystem {
  private readonly core: ArrayPathFileSystem<string, string>

  constructor(
    private readonly sep: string,
    dict: StringPathFileSystemDict = {},
  ) {
    this.core = new ArrayPathFileSystem(dict2entries(dict))
  }

  private normalize(item: string) {
    return item === '.' ? '' : item
  }

  private split(path: string) {
    return path
      .split(this.sep)
      .map(this.normalize)
      .filter(Boolean)
  }

  public existsSync = (path: string) => this.core.existsSync(this.split(path))

  public statSync = (path: string) => this.core.statSync(this.split(path))

  public readdirSync = (path: string) => this.core.readdirSync(this.split(path))

  public mkdirSync = (path: string) => this.core.mkdirSync(this.split(path))

  public readFileSync = (path: string) => this.core.readFileSync(this.split(path))

  public writeFileSync = (path: string, fileContent: string) => this.core.writeFileSync(this.split(path), fileContent)

  public ensureDirSync = (path: string) => this.core.ensureDirSync(this.split(path))

  public outputFileSync = (path: string, fileContent: string) => this.core.outputFileSync(this.split(path), fileContent)

  public ensureFileSync = (path: string) => this.outputFileSync(path, '')
}
