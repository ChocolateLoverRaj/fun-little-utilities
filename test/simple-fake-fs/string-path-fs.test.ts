import { StringPathFileSystem, ErrorKind } from 'simple-fake-fs'

function create () {
  return new StringPathFileSystem('/', {
    a: {
      0: 'zero',
      b: {
        1: 'one',
        c: {}
      }
    },
    abc: {
      def: {
        ghi: 'abcdefghi'
      }
    },
    foo: {
      bar: 'foobar',
      baz: 'foobaz'
    }
  })
}

function getError (fn: () => any) {
  let value: any = undefined

  try {
    value = fn()
  } catch (error) {
    return error
  }

  throw new Error(`Expecting function to throws an error, but it returns ${JSON.stringify(value)}`)
}

describe('constructor', () => {
  it('without dict', () => {
    expect(new StringPathFileSystem('/')).toMatchSnapshot()
  })

  it('with dict', () => {
    expect(create()).toMatchSnapshot()
  })
})

describe('existsSync', () => {
  describe('on thing that exists', () => {
    it('""', () => {
      expect(create().existsSync('')).toBe(true)
    })

    it('"."', () => {
      expect(create().existsSync('.')).toBe(true)
    })

    it('"/"', () => {
      expect(create().existsSync('/')).toBe(true)
    })

    it('"./"', () => {
      expect(create().existsSync('./')).toBe(true)
    })

    it('"/."', () => {
      expect(create().existsSync('/.')).toBe(true)
    })

    it('"a"', () => {
      expect(create().existsSync('a')).toBe(true)
    })

    it('"a/b"', () => {
      expect(create().existsSync('a/b')).toBe(true)
    })

    it('"a/b/c"', () => {
      expect(create().existsSync('a/b/c')).toBe(true)
    })

    it('"./a"', () => {
      expect(create().existsSync('./a')).toBe(true)
    })

    it('"./a/b"', () => {
      expect(create().existsSync('./a/b')).toBe(true)
    })

    it('"./a/b/c"', () => {
      expect(create().existsSync('./a/b/c')).toBe(true)
    })

    it('"a/"', () => {
      expect(create().existsSync('a/')).toBe(true)
    })

    it('"a/b/"', () => {
      expect(create().existsSync('a/b/')).toBe(true)
    })

    it('"a/b/c/"', () => {
      expect(create().existsSync('a/b/c/')).toBe(true)
    })

    it('"a/0"', () => {
      expect(create().existsSync('a/0')).toBe(true)
    })

    it('"a/b/1"', () => {
      expect(create().existsSync('a/b/1')).toBe(true)
    })

    it('"abc"', () => {
      expect(create().existsSync('abc')).toBe(true)
    })

    it('"abc/def"', () => {
      expect(create().existsSync('abc/def')).toBe(true)
    })

    it('"abc/def/ghi"', () => {
      expect(create().existsSync('abc/def/ghi')).toBe(true)
    })
  })

  describe('on thing that does not exist', () => {
    it('"Q"', () => {
      expect(create().existsSync('Q')).toBe(false)
    })

    it('"a/Q"', () => {
      expect(create().existsSync('a/Q')).toBe(false)
    })

    it('"a/b/Q"', () => {
      expect(create().existsSync('a/b/Q')).toBe(false)
    })

    it('"a/b/c/Q"', () => {
      expect(create().existsSync('a/b/c/Q')).toBe(false)
    })
  })
})

describe('statSync', () => {
  describe('on file', () => {
    describe('"a/0"', () => {
      it('has isFile() returns true', () => {
        expect(create().statSync('a/0').isFile()).toBe(true)
      })

      it('has isDirectory() returns false', () => {
        expect(create().statSync('a/0').isDirectory()).toBe(false)
      })
    })

    describe('"abc/def/ghi"', () => {
      it('has isFile() returns true', () => {
        expect(create().statSync('abc/def/ghi').isFile()).toBe(true)
      })

      it('has isDirectory() returns false', () => {
        expect(create().statSync('abc/def/ghi').isDirectory()).toBe(false)
      })
    })

    describe('"foo/bar"', () => {
      it('has isFile() returns true', () => {
        expect(create().statSync('foo/bar').isFile()).toBe(true)
      })

      it('has isDirectory() returns false', () => {
        expect(create().statSync('foo/bar').isDirectory()).toBe(false)
      })
    })

    describe('"foo/baz"', () => {
      it('has isFile() returns true', () => {
        expect(create().statSync('foo/baz').isFile()).toBe(true)
      })

      it('has isDirectory() returns false', () => {
        expect(create().statSync('foo/baz').isDirectory()).toBe(false)
      })
    })
  })

  describe('on directory', () => {
    describe('""', () => {
      it('has isFile() returns false', () => {
        expect(create().statSync('').isFile()).toBe(false)
      })

      it('has isDirectory() returns true', () => {
        expect(create().statSync('').isDirectory()).toBe(true)
      })
    })

    describe('"."', () => {
      it('has isFile() returns false', () => {
        expect(create().statSync('.').isFile()).toBe(false)
      })

      it('has isDirectory() returns true', () => {
        expect(create().statSync('.').isDirectory()).toBe(true)
      })
    })

    describe('"a"', () => {
      it('has isFile() returns false', () => {
        expect(create().statSync('a').isFile()).toBe(false)
      })

      it('has isDirectory() returns true', () => {
        expect(create().statSync('a').isDirectory()).toBe(true)
      })
    })

    describe('"a/b"', () => {
      it('has isFile() returns false', () => {
        expect(create().statSync('a/b').isFile()).toBe(false)
      })

      it('has isDirectory() returns true', () => {
        expect(create().statSync('a/b').isDirectory()).toBe(true)
      })
    })

    describe('"a/b/c"', () => {
      it('has isFile() returns false', () => {
        expect(create().statSync('a/b/c').isFile()).toBe(false)
      })

      it('has isDirectory() returns true', () => {
        expect(create().statSync('a/b/c').isDirectory()).toBe(true)
      })
    })

    describe('"abc"', () => {
      it('has isFile() returns false', () => {
        expect(create().statSync('abc').isFile()).toBe(false)
      })

      it('has isDirectory() returns true', () => {
        expect(create().statSync('abc').isDirectory()).toBe(true)
      })
    })
  })

  describe('on thing that does not exist: throws an error', () => {
    describe('"thing/that/does/not/exist"', () => {
      function init () {
        const name = 'thing/that/does/not/exist'
        const fs = create()
        const fn = () => fs.statSync(name)
        return { name, fs, fn }
      }

      it('matches snapshot', () => {
        expect(getError(init().fn)).toMatchSnapshot()
      })

      it('has .errno = ErrorKind.ENOENT', () => {
        expect(getError(init().fn)).toHaveProperty('errno', ErrorKind.ENOENT)
      })

      it('has .code = "ENOENT"', () => {
        expect(getError(init().fn)).toHaveProperty('code', 'ENOENT')
      })

      it('has .syscall = "stat"', () => {
        expect(getError(init().fn)).toHaveProperty('syscall', 'stat')
      })

      it('has .path', () => {
        const { name, fn } = init()
        expect(getError(fn)).toHaveProperty('path', name.split('/'))
      })
    })

    describe('"thing-that-does-not-exist"', () => {
      function init () {
        const name = 'thing-that-does-not-exist'
        const fs = create()
        const fn = () => fs.statSync(name)
        return { name, fs, fn }
      }

      it('matches snapshot', () => {
        expect(getError(init().fn)).toMatchSnapshot()
      })

      it('has .errno = ErrorKind.ENOENT', () => {
        expect(getError(init().fn)).toHaveProperty('errno', ErrorKind.ENOENT)
      })

      it('has .code = "ENOENT"', () => {
        expect(getError(init().fn)).toHaveProperty('code', 'ENOENT')
      })

      it('has .syscall = "stat"', () => {
        expect(getError(init().fn)).toHaveProperty('syscall', 'stat')
      })

      it('has .path', () => {
        const { name, fn } = init()
        expect(getError(fn)).toHaveProperty('path', name.split('/'))
      })
    })
  })
})

describe('readdirSync', () => {
  describe('on directory', () => {
    it('""', () => {
      expect(create().readdirSync('')).toEqual(['a', 'abc', 'foo'])
    })

    it('"a"', () => {
      expect(create().readdirSync('a')).toEqual(['0', 'b'])
    })

    it('"a/b/c"', () => {
      expect(create().readdirSync('a/b/c')).toEqual([])
    })

    it('"abc"', () => {
      expect(create().readdirSync('abc')).toEqual(['def'])
    })

    it('"abc/def"', () => {
      expect(create().readdirSync('abc/def')).toEqual(['ghi'])
    })

    it('"foo"', () => {
      expect(create().readdirSync('foo')).toEqual(['bar', 'baz'])
    })
  })

  describe('on file: throws an error', () => {
    function init () {
      const name = 'abc/def/ghi'
      const fs = create()
      const fn = () => fs.readdirSync(name)
      return { name, fs, fn }
    }

    it('matches snapshot', () => {
      expect(getError(init().fn)).toMatchSnapshot()
    })

    it('has .errno = ErrorKind.ENOTDIR', () => {
      expect(getError(init().fn)).toHaveProperty('errno', ErrorKind.ENOTDIR)
    })

    it('has .code = "ENOTDIR"', () => {
      expect(getError(init().fn)).toHaveProperty('code', 'ENOTDIR')
    })

    it('has .syscall = "scandir"', () => {
      expect(getError(init().fn)).toHaveProperty('syscall', 'scandir')
    })

    it('has .path', () => {
      const { name, fn } = init()
      expect(getError(fn)).toHaveProperty('path', name.split('/'))
    })
  })

  describe('on thing that does not exist: throws an error', () => {
    describe('"thing/that/does/not/exist"', () => {
      function init () {
        const name = 'thing/that/does/not/exist'
        const fs = create()
        const fn = () => fs.readdirSync(name)
        return { name, fs, fn }
      }

      it('matches snapshot', () => {
        expect(getError(init().fn)).toMatchSnapshot()
      })

      it('has .errno = ErrorKind.ENOENT', () => {
        expect(getError(init().fn)).toHaveProperty('errno', ErrorKind.ENOENT)
      })

      it('has .code = "ENOENT"', () => {
        expect(getError(init().fn)).toHaveProperty('code', 'ENOENT')
      })

      it('has .syscall = "scandir"', () => {
        expect(getError(init().fn)).toHaveProperty('syscall', 'scandir')
      })

      it('has .path', () => {
        const { name, fn } = init()
        expect(getError(fn)).toHaveProperty('path', name.split('/'))
      })
    })

    describe('"thing-that-does-not-exist"', () => {
      function init () {
        const name = 'thing-that-does-not-exist'
        const fs = create()
        const fn = () => fs.readdirSync(name)
        return { name, fs, fn }
      }

      it('matches snapshot', () => {
        expect(getError(init().fn)).toMatchSnapshot()
      })

      it('has .errno = ErrorKind.ENOENT', () => {
        expect(getError(init().fn)).toHaveProperty('errno', ErrorKind.ENOENT)
      })

      it('has .code = "ENOENT"', () => {
        expect(getError(init().fn)).toHaveProperty('code', 'ENOENT')
      })

      it('has .syscall = "scandir"', () => {
        expect(getError(init().fn)).toHaveProperty('syscall', 'scandir')
      })

      it('has .path', () => {
        const { name, fn } = init()
        expect(getError(fn)).toHaveProperty('path', name.split('/'))
      })
    })
  })
})

describe('readFileSync', () => {
  describe('on file', () => {
    it('"abc/def/ghi"', () => {
      expect(create().readFileSync('abc/def/ghi')).toBe('abcdefghi')
    })

    it('"foo/bar"', () => {
      expect(create().readFileSync('foo/bar')).toBe('foobar')
    })

    it('"foo/baz"', () => {
      expect(create().readFileSync('foo/baz')).toBe('foobaz')
    })
  })

  describe('on directory: throws an error', () => {
    function init () {
      const name = 'abc/def'
      const fs = create()
      const fn = () => fs.readFileSync(name)
      return { name, fs, fn }
    }

    it('matches snapshot', () => {
      expect(getError(init().fn)).toMatchSnapshot()
    })

    it('has .errno = ErrorKind.EISDIR', () => {
      expect(getError(init().fn)).toHaveProperty('errno', ErrorKind.EISDIR)
    })

    it('has .code = "EISDIR"', () => {
      expect(getError(init().fn)).toHaveProperty('code', 'EISDIR')
    })

    it('has .syscall = "read"', () => {
      expect(getError(init().fn)).toHaveProperty('syscall', 'read')
    })

    it('has .path', () => {
      const { name, fn } = init()
      expect(getError(fn)).toHaveProperty('path', name.split('/'))
    })
  })

  describe('on thing that does not exist: throws an error', () => {
    describe('"thing/that/does/not/exist"', () => {
      function init () {
        const name = 'thing/that/does/not/exist'
        const fs = create()
        const fn = () => fs.readFileSync(name)
        return { name, fs, fn }
      }

      it('matches snapshot', () => {
        expect(getError(init().fn)).toMatchSnapshot()
      })

      it('has .errno = ErrorKind.ENOENT', () => {
        expect(getError(init().fn)).toHaveProperty('errno', ErrorKind.ENOENT)
      })

      it('has .code = "ENOENT"', () => {
        expect(getError(init().fn)).toHaveProperty('code', 'ENOENT')
      })

      it('has .syscall = "read"', () => {
        expect(getError(init().fn)).toHaveProperty('syscall', 'read')
      })

      it('has .path', () => {
        const { name, fn } = init()
        expect(getError(fn)).toHaveProperty('path', name.split('/'))
      })
    })

    describe('"thing-that-does-not-exist""', () => {
      function init () {
        const name = 'thing-that-does-not-exist'
        const fs = create()
        const fn = () => fs.readFileSync(name)
        return { name, fs, fn }
      }

      it('matches snapshot', () => {
        expect(getError(init().fn)).toMatchSnapshot()
      })

      it('has .errno = ErrorKind.ENOENT', () => {
        expect(getError(init().fn)).toHaveProperty('errno', ErrorKind.ENOENT)
      })

      it('has .code = "ENOENT"', () => {
        expect(getError(init().fn)).toHaveProperty('code', 'ENOENT')
      })

      it('has .syscall = "read"', () => {
        expect(getError(init().fn)).toHaveProperty('syscall', 'read')
      })

      it('has .path', () => {
        const { name, fn } = init()
        expect(getError(fn)).toHaveProperty('path', name.split('/'))
      })
    })
  })
})

describe('mkdirSync', () => {
  describe('on name that has yet to be occupied', () => {
    function init (name: string) {
      const fs = create()
      fs.mkdirSync(name)
      return { name, fs }
    }

    describe('"abc/def/new"', () => {
      const name = 'abc/def/new'

      it('existsSync() returns true', () => {
        expect(init(name).fs.existsSync(name)).toBe(true)
      })

      it('statSync().isDirectory() returns true', () => {
        expect(init(name).fs.statSync(name).isDirectory()).toBe(true)
      })

      it('statSync().isFile() returns false', () => {
        expect(init(name).fs.statSync(name).isFile()).toBe(false)
      })

      it('readdirSync() returns an empty array', () => {
        expect(init(name).fs.readdirSync(name)).toEqual([])
      })
    })

    describe('"new"', () => {
      const name = 'new'

      it('existsSync() returns true', () => {
        expect(init(name).fs.existsSync(name)).toBe(true)
      })

      it('statSync().isDirectory() returns true', () => {
        expect(init(name).fs.statSync(name).isDirectory()).toBe(true)
      })

      it('statSync().isFile() returns false', () => {
        expect(init(name).fs.statSync(name).isFile()).toBe(false)
      })

      it('readdirSync() returns an empty array', () => {
        expect(init(name).fs.readdirSync(name)).toEqual([])
      })
    })
  })

  describe('on name that is occupied by a file: throws an error', () => {
    function init () {
      const name = 'abc/def/ghi'
      const fs = create()
      const fn = () => fs.mkdirSync(name)
      return { name, fs, fn }
    }

    it('matches snapshot', () => {
      expect(getError(init().fn)).toMatchSnapshot()
    })

    it('has .errno = ErrorKind.EEXIST', () => {
      expect(getError(init().fn)).toHaveProperty('errno', ErrorKind.EEXIST)
    })

    it('has .code = "EEXIST"', () => {
      expect(getError(init().fn)).toHaveProperty('code', 'EEXIST')
    })

    it('has .syscall = "mkdir"', () => {
      expect(getError(init().fn)).toHaveProperty('syscall', 'mkdir')
    })

    it('has .path', () => {
      const { name, fn } = init()
      expect(getError(fn)).toHaveProperty('path', name.split('/'))
    })
  })

  describe('on name that is occupied by a directory: throws an error', () => {
    function init () {
      const name = 'abc'
      const fs = create()
      const fn = () => fs.mkdirSync(name)
      return { name, fs, fn }
    }

    it('matches snapshot', () => {
      expect(getError(init().fn)).toMatchSnapshot()
    })

    it('has .errno = ErrorKind.EEXIST', () => {
      expect(getError(init().fn)).toHaveProperty('errno', ErrorKind.EEXIST)
    })

    it('has .code = "EEXIST"', () => {
      expect(getError(init().fn)).toHaveProperty('code', 'EEXIST')
    })

    it('has .syscall = "mkdir"', () => {
      expect(getError(init().fn)).toHaveProperty('syscall', 'mkdir')
    })

    it('has .path', () => {
      const { name, fn } = init()
      expect(getError(fn)).toHaveProperty('path', name.split('/'))
    })
  })

  describe('on thing that does not exist: throws an error', () => {
    function init () {
      const name = 'thing/that/does/not/exist'
      const fs = create()
      const fn = () => fs.mkdirSync(name)
      return { name, fs, fn }
    }

    it('matches snapshot', () => {
      expect(getError(init().fn)).toMatchSnapshot()
    })

    it('has .errno = ErrorKind.ENOENT', () => {
      expect(getError(init().fn)).toHaveProperty('errno', ErrorKind.ENOENT)
    })

    it('has .code = "ENOENT"', () => {
      expect(getError(init().fn)).toHaveProperty('code', 'ENOENT')
    })

    it('has .syscall = "mkdir"', () => {
      expect(getError(init().fn)).toHaveProperty('syscall', 'mkdir')
    })

    it('has .path', () => {
      const { name, fn } = init()
      expect(getError(fn)).toHaveProperty('path', name.split('/'))
    })
  })
})
