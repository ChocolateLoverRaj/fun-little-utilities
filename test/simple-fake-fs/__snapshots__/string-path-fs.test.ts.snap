// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructor with dict 1`] = `
StringPathFileSystem {
  "core": ArrayPathFileSystem {
    "coreMap": Map {
      "a" => Map {
        "0" => FakeFileContent {
          "content": "zero",
          "kind": "File",
        },
        "b" => Map {
          "1" => FakeFileContent {
            "content": "one",
            "kind": "File",
          },
          "c" => Map {},
        },
      },
      "abc" => Map {
        "def" => Map {
          "ghi" => FakeFileContent {
            "content": "abcdefghi",
            "kind": "File",
          },
        },
      },
      "foo" => Map {
        "bar" => FakeFileContent {
          "content": "foobar",
          "kind": "File",
        },
        "baz" => FakeFileContent {
          "content": "foobaz",
          "kind": "File",
        },
      },
    },
  },
  "sep": "/",
}
`;

exports[`constructor without dict 1`] = `
StringPathFileSystem {
  "core": ArrayPathFileSystem {
    "coreMap": Map {},
  },
  "sep": "/",
}
`;

exports[`ensureDirSync on child of a file: throws an error matches snapshot 1`] = `
ENOTDIR {
  "code": "ENOTDIR",
  "errno": -20,
  "path": Array [
    "abc",
    "def",
    "ghi",
    "new",
  ],
  "syscall": "mkdir",
}
`;

exports[`ensureDirSync on path that is occupied by a file: throws an error matches snapshot 1`] = `
ENOTDIR {
  "code": "ENOTDIR",
  "errno": -20,
  "path": Array [
    "abc",
    "def",
    "ghi",
  ],
  "syscall": "mkdir",
}
`;

exports[`ensureFileSync on child of a file: throws an error matches snapshot 1`] = `
ENOTDIR {
  "code": "ENOTDIR",
  "errno": -20,
  "path": Array [
    "abc",
    "def",
    "ghi",
    "new",
  ],
  "syscall": "open",
}
`;

exports[`ensureFileSync on path that is occupied by a directory: throws an error matches snapshot 1`] = `
EISDIR {
  "code": "EISDIR",
  "errno": -21,
  "path": Array [
    "abc",
    "def",
  ],
  "syscall": "open",
}
`;

exports[`mkdirSync on path that is occupied by a directory: throws an error matches snapshot 1`] = `
EEXIST {
  "code": "EEXIST",
  "errno": -17,
  "path": Array [
    "abc",
  ],
  "syscall": "mkdir",
}
`;

exports[`mkdirSync on path that is occupied by a file: throws an error matches snapshot 1`] = `
EEXIST {
  "code": "EEXIST",
  "errno": -17,
  "path": Array [
    "abc",
    "def",
    "ghi",
  ],
  "syscall": "mkdir",
}
`;

exports[`mkdirSync on thing that does not exist: throws an error matches snapshot 1`] = `
ENOENT {
  "code": "ENOENT",
  "errno": -2,
  "path": Array [
    "thing",
    "that",
    "does",
    "not",
    "exist",
  ],
  "syscall": "mkdir",
}
`;

exports[`readFileSync on directory: throws an error matches snapshot 1`] = `
EISDIR {
  "code": "EISDIR",
  "errno": -21,
  "path": Array [
    "abc",
    "def",
  ],
  "syscall": "read",
}
`;

exports[`readFileSync on thing that does not exist: throws an error "thing/that/does/not/exist" matches snapshot 1`] = `
ENOENT {
  "code": "ENOENT",
  "errno": -2,
  "path": Array [
    "thing",
    "that",
    "does",
    "not",
    "exist",
  ],
  "syscall": "read",
}
`;

exports[`readFileSync on thing that does not exist: throws an error "thing-that-does-not-exist"" matches snapshot 1`] = `
ENOENT {
  "code": "ENOENT",
  "errno": -2,
  "path": Array [
    "thing-that-does-not-exist",
  ],
  "syscall": "read",
}
`;

exports[`readdirSync on file: throws an error matches snapshot 1`] = `
ENOTDIR {
  "code": "ENOTDIR",
  "errno": -20,
  "path": Array [
    "abc",
    "def",
    "ghi",
  ],
  "syscall": "scandir",
}
`;

exports[`readdirSync on thing that does not exist: throws an error "thing/that/does/not/exist" matches snapshot 1`] = `
ENOENT {
  "code": "ENOENT",
  "errno": -2,
  "path": Array [
    "thing",
    "that",
    "does",
    "not",
    "exist",
  ],
  "syscall": "scandir",
}
`;

exports[`readdirSync on thing that does not exist: throws an error "thing-that-does-not-exist" matches snapshot 1`] = `
ENOENT {
  "code": "ENOENT",
  "errno": -2,
  "path": Array [
    "thing-that-does-not-exist",
  ],
  "syscall": "scandir",
}
`;

exports[`statSync on thing that does not exist: throws an error "thing/that/does/not/exist" matches snapshot 1`] = `
ENOENT {
  "code": "ENOENT",
  "errno": -2,
  "path": Array [
    "thing",
    "that",
    "does",
    "not",
    "exist",
  ],
  "syscall": "stat",
}
`;

exports[`statSync on thing that does not exist: throws an error "thing-that-does-not-exist" matches snapshot 1`] = `
ENOENT {
  "code": "ENOENT",
  "errno": -2,
  "path": Array [
    "thing-that-does-not-exist",
  ],
  "syscall": "stat",
}
`;

exports[`writeFileSync on path that is occupied by a directory: throws an error matches snapshot 1`] = `
EISDIR {
  "code": "EISDIR",
  "errno": -21,
  "path": Array [
    "abc",
    "def",
  ],
  "syscall": "open",
}
`;

exports[`writeFileSync on thing that does not exist: throws an error "thing/that/does/not/exist" matches snapshot 1`] = `
ENOENT {
  "code": "ENOENT",
  "errno": -2,
  "path": Array [
    "thing",
    "that",
    "does",
    "not",
    "exist",
  ],
  "syscall": "open",
}
`;
